compute_sdk/globus_compute_sdk/sdk/bash_function.py

```diff
@@ -0,0 +1,245 @@
+    def __name__(self):
+        # This is required for function registration
+        return f"{self.__class__.__name__}: {self.cmd}"
+
+    def open_std_fd(self, fname, mode: str = "a+"):
+        import os
+
+        # fname is 'stdout' or 'stderr'
+        if fname is None:
+            return None
+
+        if os.path.dirname(fname):
+            os.makedirs(os.path.dirname(fname), exist_ok=True)
+        fd = open(fname, mode)
+        return fd
+
+    def get_snippet(self, file_obj) -> str:
+        file_obj.seek(0, 0)
+        last_n_lines = file_obj.readlines()[-self.snippet_lines :]
+        return "".join(last_n_lines)
+
+    def get_and_close_streams(self, stdout, stderr) -> t.Tuple[str, str]:
+        stdout_snippet = self.get_snippet(stdout)
+        stderr_snippet = self.get_snippet(stderr)
+        stdout.close()
+        stderr.close()
+        return stdout_snippet, stderr_snippet
+
+    def execute_cmd_line(
+        self,
+        cmd: str,
+        stdout: t.Optional[str] = None,
+        stderr: t.Optional[str] = None,
+        rundir: t.Optional[str] = None,
+    ) -> BashResult:
+        import os
+        import subprocess
+        import tempfile
+
+        run_dir = rundir or self.rundir
+
+        if not run_dir:
+            # run_dir takes priority over sandboxing
+            if os.environ.get("GC_TASK_SANDBOX_DIR"):
+                run_dir = os.environ["GC_TASK_SANDBOX_DIR"]
+            else:
+                if self.run_in_sandbox and os.environ.get("GC_TASK_UUID"):
+                    run_dir = os.environ["GC_TASK_UUID"]
+                    os.environ["GC_TASK_SANDBOX_DIR"] = os.path.join(
+                        os.getcwd(), run_dir
+                    )
+
+        if run_dir:
+            os.makedirs(run_dir, exist_ok=True)
+            os.chdir(run_dir)
+
+        stdout = (
+            stdout or self.stdout or tempfile.NamedTemporaryFile(dir=os.getcwd()).name
+        )
+        stderr = (
+            stderr or self.stdout or tempfile.NamedTemporaryFile(dir=os.getcwd()).name
+        )
+
+        std_out = self.open_std_fd(stdout)
+        std_err = self.open_std_fd(stderr)
+        exception_name = None
+
+        try:
+            proc = subprocess.Popen(
+                cmd,
+                stdout=std_out,
+                stderr=std_err,
+                shell=True,
+                executable="/bin/bash",
+                close_fds=False,
+            )
+            proc.wait(timeout=self.walltime)
+            returncode = proc.returncode
+            if returncode != 0:
+                exception_name = "subprocess.CalledProcessError"
+
+        except subprocess.TimeoutExpired:
+            # Returncode to match behavior of timeout bash command
+            # https://man7.org/linux/man-pages/man1/timeout.1.html
+            returncode = 124
+            exception_name = "subprocess.TimeoutExpired"
+
+        stdout_snippet, stderr_snippet = self.get_and_close_streams(std_out, std_err)
```

Comment:
Consider calling this at the end of a try-finally clause to ensure that we close the file descriptors.

Reply:
Done.